pydantic/type_adapter.py

```diff
@@ -233,100 +208,122 @@ def __init__(
-            assert not isinstance(self._serializer, _mock_val_ser.MockValSer)
-
-    @cached_property
-    @_frame_depth(2)  # +2 for @cached_property and core_schema(self)
-    def core_schema(self) -> CoreSchema:
-        """The pydantic-core schema used to build the SchemaValidator and SchemaSerializer."""
-        if self._core_schema is None or isinstance(self._core_schema, _mock_val_ser.MockCoreSchema):
-            self._init_core_attrs(rebuild_mocks=True)  # Do not expose MockCoreSchema from public function
-        assert self._core_schema is not None and not isinstance(self._core_schema, _mock_val_ser.MockCoreSchema)
-        return self._core_schema
-
-    @cached_property
-    @_frame_depth(2)  # +2 for @cached_property + validator(self)
-    def validator(self) -> SchemaValidator | PluggableSchemaValidator:
-        """The pydantic-core SchemaValidator used to validate instances of the model."""
-        if not isinstance(self._validator, (SchemaValidator, PluggableSchemaValidator)):
-            self._init_core_attrs(rebuild_mocks=True)  # Do not expose MockValSer from public function
-        assert isinstance(self._validator, (SchemaValidator, PluggableSchemaValidator))
-        return self._validator
-
-    @cached_property
-    @_frame_depth(2)  # +2 for @cached_property + serializer(self)
-    def serializer(self) -> SchemaSerializer:
-        """The pydantic-core SchemaSerializer used to dump instances of the model."""
-        if not isinstance(self._serializer, SchemaSerializer):
-            self._init_core_attrs(rebuild_mocks=True)  # Do not expose MockValSer from public function
-        assert isinstance(self._serializer, SchemaSerializer)
-        return self._serializer
+            self.serializer = SchemaSerializer(self.core_schema, core_config)
+
+        # TODO: I think we should move this to the rebuild pattern?
+        if isinstance(self.core_schema, _mock_val_ser.MockCoreSchema):
+            self.core_schema = self.core_schema.rebuild()  # type: ignore[assignment]
+            self._init_core_attrs(ns_resolver=ns_resolver, force=True)
 
+        self._pydantic_complete = True
+        return True
+
+    @property
     def _defer_build(self) -> bool:
-        config = self._config if self._config is not None else self._model_config()
+        config = self._config if self._config is not None else self._model_config
         if config:
             return config.get('defer_build') is True
         return False
 
+    @property
     def _model_config(self) -> ConfigDict | None:
         type_: Any = _typing_extra.annotated_type(self._type) or self._type  # Eg FastAPI heavily uses Annotated
         if _utils.lenient_issubclass(type_, BaseModel):
             return type_.model_config
         return getattr(type_, '__pydantic_config__', None)
 
-    @_frame_depth(1)
+    def rebuild(
+        self,
+        *,
+        force: bool = False,
+        _parent_namespace_depth: int = 2,
```

Comment:
Where does the `2` come from?

Reply:
Good question - it's used in `parent_frame_namespace`, so that's down 1 level, then we have to add 1 more to get out of `__init__`.

So, viewed another way:

top_level
`__init__`
`parent_frame_namespace`

So depth of 2 gets us to `top_level`

Reply:
Going to resolve for now, will add a docs note explaining this